File: SecureBlackbox Samples/Password Vault/cpp/passwordvault.cpp
/*
 * SecureBlackbox 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include "../../include/secureblackbox.h"

void displayHelp() {
    printf(
        "NAME\n"
        "  passwordvault -- SecureBlackbox PasswordVault Demo Application\n\n"
        "SYNOPSIS\n"
        "  passwordvault (-set|-get|-list|-del) [-entry <entry> [-pass <password>]] [-field <field> [-value <value>]] <vault-file> [-vaultpass <vault-password>]\n\n"
        "DESCRIPTION\n"
        "  PasswordVault demonstrates the usage of PasswordVault from SecureBlackbox.\n"
        "  Used to create vaults with user information.\n\n"
        "  The options are as follows:\n\n"
        "  -set          Whether to add new entry or add/modify field value.\n\n"
        "  -get          Whether to get field value.\n\n"
        "  -list         Whether to get list of entries or fields.\n\n"
        "  -del          Whether to remove entry or field value.\n\n"
        "  -vaultpass    The password for the vault\n\n"
        "  -entry        The entry name\n\n"
        "  -pass         The password for the entry.\n\n"
        "  -field        The field name\n\n"
        "  -value        The new value of field\n\n"
        "EXAMPLES\n"
        "	passwordvault -list myvault.bin\n\n"
        "	passwordvault -get -entry myftpserver -pass mypassword -field password myvault.bin\n\n"
        "	passwordvault -set -entry myftpserver -field username -value newusername myvault.bin\n\n"
        "	passwordvault -set -entry newentry myvault.bin -vaultpass mypassword\n\n"
        "	passwordvault -del -entry myftpserver -field username myvault.bin\n\n"
    );
}

int main(int argc, char** argv) {
    PasswordVault vault;
    char* vaultfile = "";
    char* vaultpass = "";
    char* entry = "";
    char* entrypass = "";
    char* field = "";
    char* value = "";


    boolean list = false;
    boolean get = false;
    boolean set = false;
    boolean del = false;

    // Validate input
    if (argc < 3) {
        displayHelp();
        goto done;
    }

    int i = 1;
    while (i < argc) {
        if (argv[i][0] == '-') {
            if (!strcmp(argv[i], "-list")) list = true;
            else if (!strcmp(argv[i], "-get")) get = true;
            else if (!strcmp(argv[i], "-set")) set = true;
            else if (!strcmp(argv[i], "-del")) del = true;
            else if (!strcmp(argv[i], "-vaultpass")) {
                vaultpass = argv[i + 1];
                i++;
            } else if (!strcmp(argv[i], "-entry")) {
                entry = argv[i + 1];
                i++;
            } else if (!strcmp(argv[i], "-pass")) {
                entrypass = argv[i + 1];
                i++;
            } else if (!strcmp(argv[i], "-field")) {
                field = argv[i + 1];
                i++;
            } else if (!strcmp(argv[i], "-value")) {
                value = argv[i + 1];
                i++;
            }
        } else {
            vaultfile = argv[i];
        }

        i++;
    }

    if (!(list || get || set || del)) {
        printf("-list or -get or -set or -del is required.");
        displayHelp();
        goto done;
    }

    if ((list && (get || set || del)) || (get && (list || set || del)) || (set && (get || list || del)) || (del && (get ||
        set || list))) {
        printf("Use only one -list or -get or -set or -del parameter.");
        displayHelp();
        goto done;
    }

    if (!strcmp(vaultfile, "")) {
        printf("vault-file is required.");
        displayHelp();
        goto done;
    }

    vault.SetPassword(vaultpass);

    boolean openVault = false;
    const boolean saveVault = set || del;
    if (set) {
        FILE* fp = fopen(vaultfile, "rb");
        if (fp) {
            openVault = true;
            fclose(fp);
        } else {
            openVault = false;
        }
    } else {
        openVault = true;
    }

    if (openVault && vault.OpenFile(vaultfile)) {
        goto done;
    }

    if (list) {
        if (!strcmp(entry, "")) {
            char* entries = vault.ListEntries();
            if (!entries) {
                goto done;
            }

            char* res = new char[1000];
            strcpy(res, "Entries: ");
            strcat(res, entries);
            printf(res);
            delete[] res;
        } else {
            char* fields = vault.ListFields(entry, true);
            if (!fields) {
                goto done;
            }

            char* res = new char[1000];
            strcpy(res, "Fields in \"");
            strcat(res, entry);
            strcat(res, "\": ");
            strcat(res, fields);
            printf(res);
            delete[] res;
        }
    } else if (get) {
        if (!strcmp(entry, "")) {
            printf("-entry is required.");
            displayHelp();
            goto done;
        }

        if (!strcmp(field, "")) {
            printf("-field is required.");
            displayHelp();
            goto done;
        }

        vault.SetEntryPassword(entrypass);

        value = vault.GetEntryValueStr(entry, field);
        if (!value) {
            goto done;
        }

        char* res = new char[1000];
        strcpy(res, "Value: ");
        strcat(res, value);
        printf(res);
        delete[] res;
    } else if (set) {
        if (!strcmp(entry, "")) {
            printf("-entry is required.");
            displayHelp();
            goto done;
        }

        if (!strcmp(field, "")) {
            if (vault.AddEntry(entry)) {
                goto done;
            }

            char* res = new char[1000];
            strcpy(res, "Entry \"");
            strcat(res, entry);
            strcat(res, "\" successfully added.");
            printf(res);
            delete[] res;
        } else {
            vault.SetEntryPassword(entrypass);

            if (vault.SetEntryValueStr(entry, field, value, strcmp(entrypass, ""))) {
                goto done;
            }

            char* res = new char[1000];
            strcpy(res, "Field \"");
            strcat(res, field);
            strcat(res, "\" successfully added/modify.");
            printf(res);
            delete[] res;
        }
    } else {
        // del
        if (!strcmp(entry, "")) {
            printf("-entry is required.");
            displayHelp();
            goto done;
        }

        if (!strcmp(field, "")) {
            if (vault.RemoveEntry(entry)) {
                goto done;
            }

            char* res = new char[1000];
            strcpy(res, "Entry \"");
            strcat(res, entry);
            strcat(res, "\" successfully removed.");
            printf(res);
            delete[] res;
        } else {
            if (vault.RemoveField(entry, field)) {
                goto done;
            }

            char* res = new char[1000];
            strcpy(res, "Field \"");
            strcat(res, field);
            strcat(res, "\" successfully removed.");
            printf(res);
            delete[] res;
        }
    }

    if (saveVault && vault.SaveFile(vaultfile)) {
        goto done;
    }

done:
    if (vault.GetLastErrorCode()) {
        printf("Error: [%i] %s\n", vault.GetLastErrorCode(), vault.GetLastError());
    }
    getchar();
}

File: SecureBlackbox Samples/Password Vault/php/password_vault.php
<?php
/*
 * SecureBlackbox 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/secureblackbox_passwordvault.php');
require_once('../include/secureblackbox_const.php');
?>
<style>
table { width: 100% !important; }
td { white-space: nowrap; }
td input { width: 100%; }
td:last-child { width: 100%; }
</style>

<div width="90%">
  <form method=POST>
    <h2>Password vault Demo</h2>
    
    <h3>General Options</h3>
    <label><input type ="radio" checked="checked" name="comtype" value="list" /> List </label>
    <label><input type ="radio" name="comtype" value="get" /> Get </label>
    <label><input type ="radio" name="comtype" value="set" /> Set </label>
    <label><input type ="radio" name="comtype" value="del" /> Delete </label>
    <br/>
    <br/>

    <table>
      <tr><td>Vault File:</td><td><input type=text name=vaultFile value=""></td></tr>
      <tr><td>Password:</td><td><input type=password name=vaultPass value=""></td></tr>
    </table>
    <br/>
    <br/>

    <table>
      <tr><td>Entry name:</td><td><input type=text name=entryName value=""></td></tr>
      <tr><td>Entry password:</td><td><input type=password name=entryPass value=""></td></tr>
      <tr><td>Field name:</td><td><input type=text name=fieldName value=""></td></tr>
      <tr><td>Value:</td><td><input type=text name=fieldValue value=""></td></tr>
    </table>
    <br/>
    <br/>

    <input type="submit" value="Do" />
  </form>
</div><br/>

<?php
  if ($_SERVER['REQUEST_METHOD'] == "POST") {
    $vault = new SecureBlackbox_PasswordVault();
    
    try {
      // General options
      $vaultFile = $_REQUEST['vaultFile'];

      if ($vaultFile == "")
      {
        echo "<h2>Error: Please set Vault File</h2>";
        exit;
      }

      $vault->setPassword($_REQUEST['vaultPass']);

      $comtype = $_POST["comtype"];

      $openVault = False;
      $saveVault = ($comtype == "set") || ($comtype == "del");

      if ($comtype == "set")
        $openVault = file_exists($vaultFile);
      else
        $openVault = True;

      if ($openVault)
        $vault->doOpenFile($vaultFile);

      if ($comtype == "list")
      {
        if ($_REQUEST['entryName'] == "")
        {
          $entries = preg_split("/\r\n|\n|\r/", $vault->doListEntries());

          echo "<h2>Entries: </h2>";
          for($x = 0; $x < count($entries); $x++) {
            echo "<h3>" . $entries[$x] . "</h3>";
          }
        }
        else
        {
          $fields = preg_split("/\r\n|\n|\r/", $vault->doListFields($_REQUEST['entryName'], True));

          echo"<h2>Fields in " . $_REQUEST['entryName'] . ": </h2>";
          for($x = 0; $x < count($fields); $x++) {
            echo "<h3>" . $fields[$x] . "</h3>";
          }
        }
      }
      else
      if ($comtype == "get")
      {
        if ($_REQUEST['entryName'] == "")
        {
          echo "<h2>Error: Please set Entry name</h2>";
          exit;
        }

        if ($_REQUEST['fieldName'] == "")
        {
          echo "<h2>Error: Please set Field name</h2>";
          exit;
        }

        $vault->setEntryPassword($_REQUEST['entryPass']);

        $value = $vault->doGetEntryValueStr($_REQUEST['entryName'], $_REQUEST['fieldName']);
        echo "<h2>Value: </h2><p>" . $value . "</p>";
      }
      else
      if ($comtype == "set")
      {
        if ($_REQUEST['entryName'] == "")
        {
          echo "<h2>Error: Please set Entry name</h2>";
          exit;
        }

        if ($_REQUEST['fieldName'] == "")
        {
          $vault->doAddEntry($_REQUEST['entryName']);
          echo "<h2>Entry " . $_REQUEST['entryName'] . " successfully added.</h2>";
        }
        else 
        {
          try {
            $vault->doAddEntry($_REQUEST['entryName']);
          }
          catch (exception $e) {}
          $vault->setEntryPassword($_REQUEST['entryPass']);

          $vault->doSetEntryValueStr($_REQUEST['entryName'], $_REQUEST['fieldName'], $_REQUEST['fieldValue'], ($_REQUEST['entryPass'] != ""));
          echo "<h2>Field " . $_REQUEST['fieldName'] . " successfully added/modify.</h2>";
        }
      }
      else // del
      {
        if ($_REQUEST['entryName'] == "")
        {
          echo "<h2>Error: Please set Entry name</h2>";
          exit;
        }

        if ($_REQUEST['fieldName'] == "")
        {
          $vault->doRemoveEntry($_REQUEST['entryName']);
          echo "<h2>Entry " . $_REQUEST['entryName'] . " successfully removed.</h2>";
        }
        else 
        {
          $vault->doRemoveField($_REQUEST['entryName'], $_REQUEST['fieldName']);
          echo "<h2>Field " . $_REQUEST['fieldName'] . " successfully removed.</h2>";
        }
      }

      if ($saveVault)
        $vault->doSaveFile($vaultFile);
    }
    catch (exception $e) {
      echo "<h2>Failure (Details Below)</h2><p>" . $e->getMessage() . "</p>";
    }
  }
?>